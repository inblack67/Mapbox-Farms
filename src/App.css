.App{
    min-height: 100vh;
}

.input-field input{
    color: white;
}
