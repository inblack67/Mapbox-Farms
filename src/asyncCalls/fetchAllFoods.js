import axios from 'axios';
import M from 'materialize-css/dist/js/materialize.min.js';

export default async () => {
    try {
        return await axios(`https://data.ct.gov/resource/hma6-9xbg.json`);
    } catch (err) {

        if(err.response){
            M.toast({ html: `${err.response.status} Error` })
        }

        throw new Error(`Request failed with a status of ${err.response.status}`);
    }
}