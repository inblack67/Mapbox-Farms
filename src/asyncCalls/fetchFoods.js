import axios from 'axios';
import M from 'materialize-css/dist/js/materialize.min.js';

export default async ({ category, item }) => {
    try {
        return await axios(`https://data.ct.gov/resource/y6p2-px9.json?category=${category}&item=${item}`);
    } catch (err) {

        if(err.response){
            M.toast({ html: `${err.response.status} Error` })
        }

        throw new Error(`Request failed with a status of ${err.response.status}`);
    }
}