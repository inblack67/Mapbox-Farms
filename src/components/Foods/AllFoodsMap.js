import React, { useState, useContext, useEffect, useRef } from 'react'
import ReactMapGL, { Marker, Popup, FlyToInterpolator } from 'react-map-gl'
import FoodContext from '../../context/foods/foodContext'
import Preloader from '../Preloader'
import useSuperCluster from 'use-supercluster'

const AllFoodsMap = () => {

const foodContext = useContext(FoodContext);
const { allFoods, loading, getAllFoods } = foodContext;

useEffect(() => {

let mounted = true;

if(mounted){
getAllFoods();
}

return () => {
mounted = false;
}

// eslint-disable-next-line
},[])

const [viewPort, setViewPort] = useState({
latitude: 41.61609643000048,
longitude: -72.82920259099967,
width: "100vw",
height: "100vh",
zoom: 10
})

const mapRef = useRef();

const points = allFoods ? allFoods.map(food => ({
type: "Feature",
properties: { cluster: false, foodId: food.farmer_id, category: food.category,  details: { item: food.item, zipcode: food.zipcode, phone: food.phone1 } },
geometry: {
type: "Point",
coordinates: [
parseFloat(food.location_1.longitude),
parseFloat(food.location_1.latitude)
]
}
})) : []

const bounds = mapRef.current ? mapRef.current.getMap().getBounds().toArray().flat() : null; 

const { clusters, supercluster } = useSuperCluster({
points,
zoom: viewPort.zoom,
bounds,
options: {
radius: 75,
maxZoom: 20
}
})

const [selectedFarm, setSelectedFarm] = useState(null);

if(loading){
return <Preloader />
} 

return (
<div>
<ReactMapGL {...viewPort}
mapboxApiAccessToken={process.env.REACT_APP_MAPBOX_API_KEY} 
onViewportChange={viewPort => setViewPort(viewPort)}
mapStyle={process.env.REACT_APP_MAPBOX_MAP_STYLE} ref={mapRef}>

{ clusters && clusters.map(
(cluster, index) => {

const [longitude, latitude] = cluster.geometry.coordinates;

if(cluster.properties.cluster){
    return <Marker key={index} latitude={latitude} longitude={longitude}>
        <div className="container">
            <button className="btn red" onClick={ e => {
                const expansion = Math.min(supercluster.getClusterExpansionZoom(cluster.id), 20)

                setViewPort({
                    ...viewPort,
                    latitude,
                    longitude,
                    zoom: expansion,
                    transitionInterpolator: new FlyToInterpolator({
                        speed: 2
                        }),
                        transitionDuration: "auto"
                })

            } }>
                { cluster.properties.point_count }
            </button>
        </div>
    </Marker>
}

return <Marker key={index} latitude={latitude} longitude={longitude}>
<button onClick={e => setSelectedFarm(cluster)}>
    <img src="./icons/apple.svg" alt="" className="responsive-img" style={{ width: '2rem', height: '2rem' }}/>
</button>
</Marker>
}
) }

{ selectedFarm && <Popup latitude={selectedFarm.geometry.coordinates[1]} longitude={selectedFarm.geometry.coordinates[0]} onClose={ e => setSelectedFarm(null) }>
<div className="container center" style={{ width: '20rem' }}>
<p>
    <span className="red-text">Category: </span>{selectedFarm.properties.category}
</p>
<p>
    <span className="red-text">Item: </span>{selectedFarm.properties.details.item}
</p>
<p>
    <span className="red-text">Phone: </span>{selectedFarm.properties.details.phone}
</p>
<p>
    <span className="red-text">Zipcode: </span>{selectedFarm.properties.details.zipcode}
</p>
</div>
</Popup> }


</ReactMapGL>
</div>
)
}

export default AllFoodsMap;
