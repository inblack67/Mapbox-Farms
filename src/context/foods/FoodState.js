import React, { useReducer } from 'react';
import { GET_ALL_FOODS, FETCH_ERROR } from '../types';
import FoodReducer from './foodReducer';
import FoodContext from './foodContext';
import fetchAllFoods from '../../asyncCalls/fetchAllFoods'
import M from 'materialize-css/dist/js/materialize.min.js';

const FoodState = (props) => {

    const initialState = {
        allFoods: null,
        loading: true
    }

    const [state, dispatch] = useReducer(FoodReducer, initialState);

    const getAllFoods = async () => {
        try {
            const res = await fetchAllFoods();

            const data = res.data;

            if(data.length === 0){
                M.toast({ html: 'Nothing Found' });
            }

            dispatch({
                type: GET_ALL_FOODS,
                payload: data
            });
            
        } catch (err) {
            if(err.response){
                M.toast({ html: `${err.response.status} Error` })
            }
            console.error(err);
            dispatch({
                type: FETCH_ERROR
            })
        }
    }

    return (
        <FoodContext.Provider value={{
            allFoods: state.allFoods,
            getAllFoods
        }}>
            { props.children }
        </FoodContext.Provider>
    )
}

export default FoodState;
