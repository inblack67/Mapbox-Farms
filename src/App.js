import React, { useEffect } from 'react';
import './App.css';
import 'materialize-css/dist/css/materialize.min.css';
import M from 'materialize-css/dist/js/materialize.min.js';
import FoodState from './context/foods/FoodState'
import SearchFoods from './components/Foods/SearchFoods';
import FoodMap from './components/Foods/FoodMap'

function App() {

  useEffect(() => {
    // Materialize JS init
    M.AutoInit();
  });

  return (
    <FoodState>
      <div className="App black">
        <FoodMap />
        <SearchFoods />
      </div>
    </FoodState>
  );
}

export default App;
