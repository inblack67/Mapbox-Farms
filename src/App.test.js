import React from 'react';
import { render, screen } from '@testing-library/react';
import App from './App';
import AllFoodsMap from './components/Foods/AllFoodsMap'
import FoodState from './context/foods/FoodState'
import fetchFoods from './asyncCalls/fetchFoods';
import fetchAllFoods from './asyncCalls/fetchAllFoods';

const renderWithContext = (component) => {
    return render(<FoodState>
        {component}
    </FoodState>)
}

it('renders App component correctly', () => {
    const {asFragment} = render(<App />);
    expect(asFragment).toMatchSnapshot();
});

it('renders FoodMap component correctly', () => {
    const {asFragment} = renderWithContext(<AllFoodsMap />);

    expect(asFragment).toMatchSnapshot();
});

it('fetches all food trucks/farms', async () => {
    const res = await fetchAllFoods();

    expect(res.data[0].category).toEqual('Farm Products');

    expect(res.data[0].location_1).toEqual( { longitude: '-71.98244322', latitude: '41.3253323' });
})

it('fetches searched food trucks/farms', async () => {
    const res = await fetchFoods({ category: 'Fruit', item: 'Peaches' });

    expect(res.data[0].category).toEqual('Fruit');

    expect(res.data[0].location_1).toEqual({ type: 'Point', coordinates: [-79.402352, 45.649393] });
})