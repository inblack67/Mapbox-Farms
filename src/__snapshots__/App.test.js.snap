// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App component correctly 1`] = `[Function]`;

exports[`renders FoodMap component correctly 1`] = `[Function]`;
